rpc: refactor: use string_view in Arg/MaybeArg

stickies-v opened the pull request:
The `RPCHelpMan::{Arg,MaybeArg}` helpers avoid copying (potentially) large strings by returning them as `const std::string*` (`MaybeArg`) or `const std::string&` (`Arg`). For `MaybeArg`, this has the not-so-nice effect that users need to deal with raw pointers, potentially also requiring new functions (e.g. [`EnsureUniqueWalletName` ](https://github.com/bitcoin/bitcoin/pull/32845/commits/d127b25199118756122232d9aafff19f1922869b#diff-d8bfcfbdd5fa7d5c52d38c1fe5eeac9ce5c5a794cdfaf683585140fa70a32374R32)) with raw pointers being implemented.

This PR aims to improve on this by returning a trivially copyable `std::string_view` (`Arg`) or `std::optional<std::string_view>` (`MaybeArg`), modernizing the interface without introducing any additional copying overhead. In doing so, it also generalizes whether we return by value or by pointer/reference using `std::is_trivially_copyable_v` instead of defining the types manually.

In cases where functions currently take a `const std::string&` and it would be too much work / touching consensus logic to update them (`signmessage.cpp`), a `std::string` copy is made (which was already happening anyway).

The last 2 commits increase usage of the `{Arg,MaybeArg}<std::string_view>` helpers, and could be dropped/pruned if anything turns out to be controversial - I just think it's a nice little cleanup.

stickies-v committed:
rpc: refactor: use string_view in Arg/MaybeArg

Modernizes interface by not forcing users to deal with raw pointers,
without adding copying overhead. Generalizes the logic of whether
we return by value or by optional/pointer.

In cases where functions take a `const std::string&` and it would
be too much work to update them, a string copy is made (which was
already happening anyway).

stickies-v committed:
refactor: increase string_view usage

Update select functions that take a const std::string& to take a
std::string_view instead. In a next commit, this allows us to use
the {Arg,MaybeArg}<std::string_view> helper.

stickies-v committed:
rpc: refactor: use more (Maybe)Arg<std::string_view>

Use the {Arg,MaybeArg}<std::string_view> helper in all places where
it is a trivial change. In many places, this simplifies the logic
and reduces duplication of default values.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32983/commits/319abd609b3d2194e7a05454b6af55b6f16adbf1

pablomartin4btc reviewed (commented):
ACK 319abd609b3d2194e7a05454b6af55b6f16adbf1

I like this refactoring  — the general idea of reducing unnecessary `string` allocations via `string_view` makes sense.

I’m a bit unsure about the changes in `mining.cpp` and `signmessage.cpp`. It seems like we’re going from `string` → `string_view` → `string`, which introduces a bit of unnecessary churn and could lead to confusion down the line. I understand the intention may be to eventually update those functions to accept `string_view`(?), but without that broader change in place, it’s hard to see a clear benefit to doing it this way for now.

pablomartin4btc reviewed (commented):
re-ACK 122b432b71e148cd85edcd74df741f99ac7f187c

Not blocking: I’d personally prefer keeping `Arg<std::string>` in `signmessage.cpp` if we're not planning to update `MessageVerify` to accept `string_view` instead of `const std::string&`. Is updating `MessageVerify` something you’d prefer to avoid at this stage?

stickies-v commented:
Force-pushed from [122b432](https://github.com/bitcoin/bitcoin/commit/122b432b71e148cd85edcd74df741f99ac7f187c) to [137319b](https://github.com/bitcoin/bitcoin/commit/137319b5d6b0c4fa9e015608d6b43018f08db45c) to address silent merge conflict from https://github.com/bitcoin/bitcoin/pull/32845:
- replaced `self.MaybeArg<std::string>` with `self.MaybeArg<std::string_view>` in `wallet/rpc/wallet.cpp`
- updated `EnsureUniqueWalletName` signature to take `std::optional<std::string_view>` instead of `const std::string*`.

Otherwise no changes.

maflcko reviewed (approved):
Looks like a nice change using more named `Arg` and `MaybeArg` helpers, removing the manual `params[idx].isNull` and `params[idx].get_str` legacy handling.

Not a blocker, but it would be nice to fix the theoretical c_str/data issue.

lgtm ACK 24d48638795ba61e8a420df4b831281455e208a7 🏉

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK 24d48638795ba61e8a420df4b831281455e208a7 🏉
5H7u/Rl/t/WytDb3+AuCk6ztGaKQRbxTrL0FpxHFBTYItwNTR1lUNNpUXBYo9owTgz4R28dgoi482SWydKgABw==
```

</details>

stickies-v reviewed (commented):
Force-pushed to address review comments from @maflcko:
- fixed docstring 
- removed usage of std::string_view::data() to avoid null termination issues in the future
- added one more usage of Arg<string_view> helper

pablomartin4btc reviewed (approved):
crACK & tACK b49a4f17aba76d2f2d7f1109d7e68b02303947bf

Verified addressed comments & suggestions from @maflcko since my last [review](https://github.com/bitcoin/bitcoin/pull/32983#pullrequestreview-3051797902).

Performed some manual testing on RPC calls (`getblock`, `addnode`, `getdescriptoractivity`, `signmessage`, `verifymessage`). Also tested them passing very long strings (>10MB).

Left a couple of tiny nits, none blocker.

stickies-v reviewed (commented):
Force-pushed to address merge conflict from #32326, and addressed 2 review nits from @pablomartin4btc :
- replaced `strprintf` usage with `tfm::format`
- replaced `descriptor.size() == 0` with `descriptor.empty()`

pablomartin4btc reviewed (approved):
re-ACK [b63428a](https://github.com/bitcoin/bitcoin/commit/b63428ac9ce2c903670409b3e47b9f6730917ae8)

_(tiny nit - sorry, just realised this - only if you have to retouch or rebase again, it would be good to update the copyrights on the files changed that requires it)_

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/32983/commits/b63428ac9ce2c903670409b3e47b9f6730917ae8

Sjors commented:
See #33914 for a followup.
